src/file/properties.rs

```diff
@@ -0,0 +1,550 @@
+    self
+  }
+
+  // ----------------------------------------------------------------------
+  // Setters for a specific column
+
+  /// Helper method to get existing or new mutable reference of column properties.
+  #[inline]
+  fn get_mut_props(&mut self, col: ColumnPath) -> &mut ColumnProperties {
+    self.column_properties.entry(col).or_insert(ColumnProperties::new())
+  }
+
+  /// Sets encoding for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_encoding(mut self, col: ColumnPath, value: Encoding) -> Self {
+    self.get_mut_props(col).set_encoding(value);
+    self
+  }
+
+  /// Sets compression codec for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_compression(mut self, col: ColumnPath, value: Compression) -> Self {
+    self.get_mut_props(col).set_compression(value);
+    self
+  }
+
+  /// Sets flag to enable/disable dictionary encoding for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_dictionary_enabled(mut self, col: ColumnPath, value: bool) -> Self {
+    self.get_mut_props(col).set_dictionary_enabled(value);
+    self
+  }
+
+  /// Sets flag to enable/disable statistics for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_statistics_enabled(mut self, col: ColumnPath, value: bool) -> Self {
+    self.get_mut_props(col).set_statistics_enabled(value);
+    self
+  }
+
+  /// Sets max size for statistics for a column.
+  /// Takes precedence over globally defined settings.
+  pub fn set_column_max_statistics_size(mut self, col: ColumnPath, value: usize) -> Self {
+    self.get_mut_props(col).set_max_statistics_size(value);
+    self
+  }
+}
+
+/// Container for column properties that can be changed as part of writer.
+///
+/// If a field is `None`, it means that no specific value has been set for this column,
+/// so some default value must be used.
+#[derive(Debug, Clone, PartialEq)]
+struct ColumnProperties {
+  encoding: Option<Encoding>,
+  codec: Option<Compression>,
+  dictionary_enabled: Option<bool>,
+  statistics_enabled: Option<bool>,
+  max_statistics_size: Option<usize>
+}
+
+impl ColumnProperties {
+  /// Initialise column properties with default values.
+  fn new() -> Self {
+    Self {
+      encoding: None,
+      codec: None,
+      dictionary_enabled: None,
+      statistics_enabled: None,
+      max_statistics_size: None
+    }
+  }
+
+  /// Sets encoding for this column.
+  ///
+  /// This field is treated as fallback encoding, and should not be dictionary encoding.
+  /// For that there is an additional boolean flag provided.
+  fn set_encoding(&mut self, value: Encoding) {
+    if value == Encoding::PLAIN_DICTIONARY || value == Encoding::RLE_DICTIONARY {
+      panic!("Dictionary encoding can not be used as fallback encoding");
+    }
+    // TODO: Check if we need to force PLAIN encoding as dictionary fallback.
```

Comment:
This is an interesting question. Parquet-cpp forces PLAIN encoding as fallback, but I could not find any mentioning of having fallback as PLAIN only - I guess, technically any encoding could be used for fallback, except dictionary encoding. Right now it is possible to set any encoding for fallback.

Reply:
I see. I think parquet-cpp supports other types of decoding but haven't used them in the writer path, so currently they only support PLAIN in fallback as a side effect. In parquet-mr though, other types of fallback encoding schemas are used so I think we are on the correct track.